### PR TITLE
[UI] [Nodes] Disable actions for dead nodes

### LINKED ISSUE #32593: [dashboard] Logs button shows for dead nodes but doesn't work
Repro: start multi-node cluster, kill node.

Screen Shot 2023-02-15 at 5 48 58 PM

### PR DESCRIPTION
**Pull request by chaowanggg**
## Why are these changes needed?
We are going to remove logs for dead nodes and remove Stack Trace and CPU Flame Graph for dead workers.
Otherwise, users will click the link and jump to a page without information they need.

Alive Nodes | Dead Nodes
-- | --
image |  image

## Summary of Changes
1. hide logs for dead nodes in NodeRow.tsx

## Discussion and Confusion
As part of the PR, we have also come to an agreement, which I will list here for future reference.

1. There are three entities on the Cluster page: node, raylet, and worker.
2. We have a state field for node and raylet, but we do not have the state of the worker returned from the backend API.
3. In Ray, a node is equal to a raylet.

## Related issue number
fixes #32593

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by chaowanggg** [comment]
needs more discussion with @scottsun94 about how to deal with dead workers logs

File: dashboard/client/src/pages/node/NodeRow.tsx [hunk 1]
```diff
@@ -195,28 +198,33 @@ export const WorkerRow = ({ node, worker }: WorkerRowProps) => {
       </TableCell>
       <TableCell align="center">{pid}</TableCell>
       <TableCell>
+        {/* {state !== "DEAD" && ( */}
```
Comment (alanwguo) [inline dashboard/client/src/pages/node/NodeRow.tsx:201]: remove

File: dashboard/client/src/pages/node/NodeRow.tsx [hunk 2]
```diff
@@ -195,28 +198,33 @@ export const WorkerRow = ({ node, worker }: WorkerRowProps) => {
...
-        <a
-          href={`/worker/cpu_profile?pid=${pid}&ip=${ip}&duration=5&native=0`}
-          target="_blank"
-          title="Profile the Python worker for 5 seconds (default) and display a CPU flame graph."
-          rel="noreferrer"
-        >
-          CPU&nbsp;Flame&nbsp;Graph
-        </a>
-        <br />
+        {/* )} */}
```
Comment (alanwguo) [inline dashboard/client/src/pages/node/NodeRow.tsx:205]: remove

File: dashboard/client/src/pages/node/NodeRow.component.test.tsx [hunk 3]
```diff
@@ -153,4 +141,27 @@ describe("WorkerRow", () => {
     // Memory Usage
     expect(screen.getByText(/75\.0000B\/100\.0000B\(75\.0%\)/)).toBeVisible();
   });
+  it("Disable CPU profiling & stacktrace for dead worker", async () => {
```
Comment (alanwguo) [inline dashboard/client/src/pages/node/NodeRow.component.test.tsx:144]: nit: add newline between test cases

File: dashboard/client/src/pages/node/NodeRow.component.test.tsx [hunk 4]
```diff
@@ -77,64 +99,30 @@ describe("NodeRow", () => {
     expect(screen.getByText(/5.0000B\/s/)).toBeVisible();
     expect(screen.getByText(/10.0000B\/s/)).toBeVisible();
   });
+  it("Disable actions for Dead node", async () => {
```
Comment (alanwguo) [inline dashboard/client/src/pages/node/NodeRow.component.test.tsx:102]: nit: add newlinen between test cases

**Review by alanwguo — COMMENTED** [review]
For all UI changes, please add a screenshot to the PR description.

File: dashboard/client/src/pages/node/NodeRow.tsx (hunk 1 above)
Comment (chaowanggg) [reply dashboard/client/src/pages/node/NodeRow.tsx:201]: Yes, we should not leave unnecessary comment in the code. I will make the changes after getting the UI decision from @scottsun94 to decide whether to show the logs of dead worker/nodes

**Comment by chaowanggg** [comment]
@alanwguo I am curious why do hardcode the status of a worker as "ALIVE"? 

For this task, I need to hide  CPU profiling & stack trace for DEAD workers and I think I could judge whether it is dead by using "is_alive" prop from the [API](https://docs.ray.io/en/latest/ray-observability/api/state/doc/ray.experimental.state.common.WorkerState.html#ray.experimental.state.common.WorkerState)

https://github.com/ray-project/ray/blob/50aebb861c2ddf11f520024aa5b059b28a8ef38a/dashboard/client/src/pages/node/NodeRow.tsx#L190

**Comment by alanwguo** [comment]
> @alanwguo I am curious why do we hardcode the status of a worker as "ALIVE"?
> 
> For this task, I need to hide CPU profiling & stack trace for DEAD workers and I think I could judge whether it is dead by using "is_alive" prop from the [API](https://docs.ray.io/en/latest/ray-observability/api/state/doc/ray.experimental.state.common.WorkerState.html#ray.experimental.state.common.WorkerState)
> 
> https://github.com/ray-project/ray/blob/50aebb861c2ddf11f520024aa5b059b28a8ef38a/dashboard/client/src/pages/node/NodeRow.tsx#L190

Hm. I'm not sure. I think at one point we only showed alive Nodes in the list view but that is no longer the case. Please remove the hard-coding.

**Comment by chaowanggg** [comment]
> > @alanwguo I am curious why do we hardcode the status of a worker as "ALIVE"?
> > For this task, I need to hide CPU profiling & stack trace for DEAD workers and I think I could judge whether it is dead by using "is_alive" prop from the [API](https://docs.ray.io/en/latest/ray-observability/api/state/doc/ray.experimental.state.common.WorkerState.html#ray.experimental.state.common.WorkerState)
> > https://github.com/ray-project/ray/blob/50aebb861c2ddf11f520024aa5b059b28a8ef38a/dashboard/client/src/pages/node/NodeRow.tsx#L190
> 
> Hm. I'm not sure. I think at one point we only showed alive Nodes in the list view but that is no longer the case. Please remove the hard-coding.

Thanks Alan.  I just wanted to double-confirm that the backend logic has filtered the workers and is only returning "ALIVE" workers to the front end. This is because I need to display different UI for workers that are "ALIVE" versus those that are "DEAD".

Could you please let me know whom I should contact to confirm the backend logic, or provide me with a code pointer that I can refer to?

**Comment by alanwguo** [comment]
Please check `get_all_nodes` in `node_head.py`

**Comment by chaowanggg** [comment]
@alanwguo In this PR, I need to demonstrate distinct UIs for dead workers and alive workers. However, I'm unable to determine the state of a worker(not node) using the available API, as it doesn't include a "state" or "is_alive" field

image

**Comment by chaowanggg** [comment]
> Please check `get_all_nodes` in `node_head.py`

Double confirmed, no filtering in `get_all_nodes` function

**Comment by alanwguo** [comment]
> @alanwguo In this PR, I need to demonstrate distinct UIs for dead workers and alive workers. However, I'm unable to determine the state of a worker(not node) using the available API, as it doesn't include a "state" or "is_alive" field
> 
> image

I think all workers are guaranteed to be alive. We don't show information about dead workers.

Can we apply this change only to nodes for this PR?

**Comment by chaowanggg** [comment]
@scottsun94 All workers are guaranteed to be alive and we don't show information about dead workers. 

According to this [figma](https://www.figma.com/file/1BaUL5zyv6rlnOo47Iio4u/%5Bvision%5D-Anyscale%2FRay-Observability?node-id=1209-83314&t=3YNPNV87iS0rnTk6-0). Do you think we could skip the UI changes for workers?

**Comment by scottsun94** [comment]
> @scottsun94 All workers are guaranteed to be alive and we don't show information about dead workers.
> 
> According to this [figma](https://www.figma.com/file/1BaUL5zyv6rlnOo47Iio4u/%5Bvision%5D-Anyscale%2FRay-Observability?node-id=1209-83314&t=3YNPNV87iS0rnTk6-0). Do you think we could skip the UI changes for workers?

SG. Let's skip it for this PR.


However, is there any reason why we don't expose/have the info about the dead workers? @rkooo567 
Let's say if a worker died because of OOMs, and usually the error message asks people to look at the worker logs. If a dead worker is gone, they have to go to the logs UI to find the logs. With the new worker log change, it's really hard for people to find it any more...

File: dashboard/client/src/pages/node/NodeRow.tsx [hunk 5]
```diff
@@ -94,7 +95,9 @@ export const NodeRow = ({
         </Box>
       </TableCell>
       <TableCell>
-        <Link to={`/logs/${encodeURIComponent(logUrl)}`}>Log</Link>
```
Comment (chaowanggg) [inline dashboard/client/src/pages/node/NodeRow.tsx:97]: I tried to get the state of a node to decide whether to show logs or hide logs. But I noticed that in line 78, we are using raylet.state. 
<TableCell>
        <StatusChip type="node" status={raylet.state} />
</TableCell>

Thus, we have two state here, node's state and raylet's state.

Why do we need to show raylet's state here? Do I need to hide or show the logs based on raylet's state?
@alanwguo

Comment (alanwguo) [reply dashboard/client/src/pages/node/NodeRow.tsx:97]: @rkooo567 can you explain the difference between these states?

Comment (rkooo567) [reply dashboard/client/src/pages/node/NodeRow.tsx:97]: node == raylet in Ray

Comment (rkooo567) [reply dashboard/client/src/pages/node/NodeRow.tsx:97]: Besides, we can refactor these API responses from the dashboard at some point..

**Review by rkooo567 — APPROVED** [review]

**Comment by rkooo567** [comment]
oops sorry ignore my aproval ^

File: dashboard/client/src/pages/node/NodeRow.tsx (hunk 5 above)
Comment (chaowanggg) [reply dashboard/client/src/pages/node/NodeRow.tsx:97]: Good idea!

**Review by alanwguo — APPROVED** [review]

**Pull request merged**